### ISSUE TITLE
Get-BCArtifactUrl is very slow (mainly for Sandbox artifacts)

### ISSUE THREAD

**Issue by Koubek** [microsoft/navcontainerhelper#3268]
PLEASE DO NOT INCLUDE ANY PASSWORDS OR TOKENS IN YOUR ISSUE!!!

**Describe the issue**
Command `Get-BCArtifactUrl -Type Sandbox` takes very long to process. It happens mainly with **Sandbox** artifacts. I assume it's because there is a huge list of the artifacts behind the scene. Or maybe not, I can't tell :) But if yes, the question is if all of them are needed.

**Scripts used to create container and cause the issue**
```
Get-BCArtifactUrl -Type Sandbox
Get-BCArtifactUrl -Type Sandbox -select Current
# ... and similar ...
```

**Additional context**
I think the time needed to process the request grows over time but I haven't made any serious measurements.
I assume some cleanup could help probably to make the query perform faster ;)

**Comment by freddydk** [comment]
You are right - it is caused by the very large list of artifacts - we do cleanup the list occasionally - will see what I can do

**Comment by freddydk** [comment]
The "real" for for this however is when we start using NuGet, Universal packages or like for artifacts instead of blob storage - that is on the horizon - but not anytime soon.

**Comment by Koubek** [comment]
I know this might be off-topic, but have you already checked the performance on such a big volume of data e.g. with NuGet? Partially, it's just a curiosity and partially not - we use NuGet and have a good experience performance-wise so far but this might change over time so someone with a significant volume of the data would be a good example for us (positive and eventually also negative).

**Comment by tfenster** [comment]
@Koubek if you don't mind a bit of caching in between, you could use https://bca-url-proxy.azurewebsites.net/bca-url/sandbox/de/?donotredirect=true to get the URL. It is as slow as bcch (which it uses behind the scenes) on the first request for a particular artifact, but then it is fast as it takes the result from a cache. After 1h for sandbox artifactss and 24h for onprem artifacts, the cache entry expires and fetches the URL through bcch again. So depending on your usage scenario, this might at least help :)

**Comment by rdebath** [comment]
I notice that current versions of the BLOB storage can assign tags to the images.
These tags can be searched on and apparently changed without re-uploading the blob.

Tags like "country", "major_version"  would be obvious filter candidates.
Ones like "latest", "previous","recent" could be attached to provide (hopefully) super quick searches for the related `bcSelect` values.

**Comment by freddydk** [comment]
I have done a lot of investigations into this - and we will come up with a solution for this (both search and download speed) - we will share a design in a few weeks.

**Comment by BW-PA** [comment]
Hi Freddy, do you have any updates to share on this matter?  I came here looking as a build I was monitoring today on one of our self hosted Azure DevOps agents took 2min 51secs (of an elapsed 5min 59secs total) waiting for the "Determining artifacts to use" statement to complete. 

Here's the statement from the raw log in DevOps
```
2024-10-22T05:00:14.9283994Z Determining artifacts to use
2024-10-22T05:03:06.6964906Z ##[group]Parameters
```

**Comment by freddydk** [comment]
Unfortunately not.
Using the public web service above can speed things up quite a bit

**Comment by joandrsn** [comment]
Hi @freddydk,

I’d like to follow up on this issue.

I see that you've attempted to enforce the use of a version number, but since it's opt-in, it hopefully doesn't break existing behavior:
[Relevant Code](https://github.com/microsoft/navcontainerhelper/blob/5828402c34669155ac99e15187164d349af87952/Artifacts/Get-BCArtifactUrl.ps1#L145-L183)

This works well, but since you also consider it a hack, perhaps we can find a better solution.
# The Problem: Too Many Blobs

As mentioned before, there are currently **331,245 blobs** in the Sandbox container (measured with `Get-BCArtifactUrl -type Sandbox -select All | Measure-Object`).
Even filtering by version doesn’t fully solve the issue-for example, BC24 alone has **114,159 blobs**:
```powershell
Get-BCArtifactUrl -type Sandbox -select All -version 24 | Measure-Object
 ```
We need better filtering when dealing with such large numbers.
# Proposal: Use Tags for Filtering

As @rdebath suggested, [using tags ](https://learn.microsoft.com/en-us/rest/api/storageservices/find-blobs-by-tags?tabs=microsoft-entra-id)would allow for server-side filtering, significantly reducing the number of blobs listed. However, this would require Microsoft to update the tags as part of their pipeline (e.g., via a PowerShell script).

Some possible filtering approaches:
## 1. Country Filter (Massive Reduction)

Filtering by country alone could cut down blobs by ~50x:

```powershell
Get-BCArtifactURL -country dk
```
This would result in a filter like:
```
WHERE=country = 'dk'
``` 
## 2. Version Filter

Using tags for versioning would still be efficient (though no prefix filtering on the tags endpoint):
```powershell
Get-BCArtifactURL -version 25.4
```
Equivalent filter:
```
WHERE major = 25 AND minor = 4
```
## 3. Combined Filtering (Best Case)

Combining both country and version filters would be even better:

```powershell
Get-BCArtifactURL -version 25.4 -country dk
```
```
WHERE major = 25 AND minor = 4 AND country = 'dk'
```
## 4. Default Filtering When No Hints Are Provided

Without hints, listing blobs remains slow:
```powershell
Get-BCArtifactURL
```
A possible solution:
```
WHERE major >= 25
```
The threshold (25) could be hardcoded or configurable in `bccontainerhelper`'s config.
This would reduce the list to ~33,100 blobs instead of 330,000+.

I would love to hear your thoughts on this approach.

**Comment by freddydk** [comment]
What you can do today (to speed up getting the latest version) is to use

`$bcContainerHelperConfig.useApproximateVersion =  $true`

which will reduce the time for getting an artifact url to a few seconds.

but... - a few things...

We are working on moving artifacts away from the blob store and to OCI artifacts.
During this process, we will:
- Create an index for all artifacts per country
- Publish fewer artifacts (only artifacts actually published online)
- Use a more aggressive retention policy

**Comment by rdebath** [comment]
I don't see the advantage of putting windows container images into an OCI wrapper rather than a Docker one.  But something with better indexing than the current blob store would be a very good thing.

I don't think "fewer artifacts" is the way to go, however.

Unless this is a way of splitting binaries from source. I for one have only limited interest in the continual compiles of the "platform"; specifically, recent master versions and specific versions to match other "published online"; sandbox and onprem versions.

But I have a git repository with the AL source of nearly every version, this is very useful for tracking (breaking) changes. It's also not that large, 5.5G for everything (onprem and SaaS) back to V15. (I skipped V12) I would not want to lose that information now that I have it.

NB: I also have C/AL history back a long way.

**Comment by freddydk** [comment]
@rdebath - we are not talking about container images, I am talking about the blobs.
Downloading 2Gb from OCI artifacts is much faster than from blobs.
But... - it all comes down to what we are allowed to do internally in Microsoft, we are discussing that and finding the best option.
